include/boost/math/statistics/detail/single_pass.hpp

```diff
@@ -0,0 +1,384 @@
+//  (C) Copyright Nick Thompson 2018
+//  (C) Copyright Matt Borland 2020
+//  Use, modification and distribution are subject to the
+//  Boost Software License, Version 1.0. (See accompanying file
+//  LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef BOOST_MATH_STATISTICS_UNIVARIATE_STATISTICS_DETAIL_SINGLE_PASS_HPP
+#define BOOST_MATH_STATISTICS_UNIVARIATE_STATISTICS_DETAIL_SINGLE_PASS_HPP
+
+#include <tuple>
+#include <iterator>
+#include <atomic>
+#include <memory>
+#include <thread>
+#include <type_traits>
+#include <future>
+#include <cmath>
+#include <algorithm>
+#include <valarray>
+#include <mutex>
+#include <unordered_map>
+
+namespace boost::math::statistics::detail
+{
+template<typename ReturnType, typename ForwardIterator>
+ReturnType mean_sequential_impl(ForwardIterator first, ForwardIterator last)
+{
+    if constexpr (std::is_same_v<typename std::iterator_traits<ForwardIterator>::iterator_category, std::random_access_iterator_tag>)
+    {
+        std::size_t elements {std::distance(first, last)};
+        std::valarray<ReturnType> mu {0, 0, 0, 0};
+        std::valarray<ReturnType> temp {0, 0, 0, 0};
+        ReturnType i = 1;
+        const auto end {last - (elements % 4)};
+
+        for(auto it {first}; it != end; it += 4)
+        {
+            const ReturnType inv {ReturnType(1) / i};
+            temp = {*it, *(it+1), *(it+2), *(it+3)};
+            temp -= mu;
+            mu += (temp *= inv);
+            i += 1;
+        }
+
+        const ReturnType num1 = ReturnType(elements - (elements %4))/ReturnType(4);
+        const ReturnType num2 = num1 + ReturnType(elements % 4);
+
+        for (auto it = end; it != last; ++it)
+        {
+            mu[3] += (*it-mu[3])/i;
+            i += 1;
+        }
+
+        return (num1 * std::valarray<ReturnType>(mu[std::slice(0,3,1)]).sum() + num2 * mu[3]) / ReturnType(elements);
```

Comment:
This should reference the appropriate equation in the reference for parallel reduction of means.